File: uniconn-go/secure/handshake.go
package secure

import (
	"crypto/mlkem"
	"fmt"
	"io"

	"lukechampine.com/blake3"
)

// sessionKeys holds the derived encryption keys and nonce prefixes.
type sessionKeys struct {
	InitiatorKey      [32]byte
	ResponderKey      [32]byte
	InitiatorNoncePfx [NoncePrefixSize]byte
	ResponderNoncePfx [NoncePrefixSize]byte
}

// deriveKeys derives session keys from the ML-KEM shared secret using BLAKE3.
func deriveKeys(sharedSecret []byte) *sessionKeys {
	h := blake3.New(KDFOutputSize, nil)
	h.Write([]byte(KDFContext))
	h.Write(sharedSecret)
	out := h.Sum(nil)

	sk := &sessionKeys{}
	copy(sk.InitiatorKey[:], out[0:32])
	copy(sk.ResponderKey[:], out[32:64])
	copy(sk.InitiatorNoncePfx[:], out[64:76])
	copy(sk.ResponderNoncePfx[:], out[76:88])
	return sk
}

// HandshakeInitiator performs the initiator side of the USCP handshake.
//
//  1. Generate ephemeral ML-KEM-1024 key pair
//  2. Sign the encapsulation key with our ML-DSA-87 identity
//  3. Send HELLO
//  4. Receive HELLO_REPLY
//  5. Verify peer fingerprint and signature
//  6. Decapsulate shared secret
//  7. Derive session keys
func HandshakeInitiator(rw io.ReadWriter, id *Identity, peerFP Fingerprint) (*SecureConn, error) {
	// 1. Ephemeral ML-KEM-1024.
	dk, err := mlkem.GenerateKey1024()
	if err != nil {
		return nil, fmt.Errorf("mlkem keygen: %w", err)
	}
	ek := dk.EncapsulationKey().Bytes()

	// 2. Sign ek.
	sig, err := id.Sign(ek)
	if err != nil {
		return nil, fmt.Errorf("sign ek: %w", err)
	}

	// 3. Send HELLO.
	pubBytes, _ := id.PublicKey.MarshalBinary()
	hello := &helloMsg{
		Type:      MsgHello,
		PublicKey: pubBytes,
		Payload:   ek,
		Signature: sig,
	}
	if _, err := rw.Write(hello.marshal()); err != nil {
		return nil, fmt.Errorf("send hello: %w", err)
	}

	// 4. Receive HELLO_REPLY.
	reply, err := unmarshalHello(rw.(io.Reader))
	if err != nil {
		return nil, fmt.Errorf("recv hello_reply: %w", err)
	}
	if reply.Type != MsgHelloReply {
		return nil, fmt.Errorf("expected HELLO_REPLY (0x%02x), got 0x%02x", MsgHelloReply, reply.Type)
	}

	// 5. Verify peer.
	peerPub, err := parseMDSA87PublicKey(reply.PublicKey)
	if err != nil {
		return nil, err
	}
	gotFP := computeFingerprintBytes(reply.PublicKey)
	if gotFP != peerFP {
		return nil, fmt.Errorf("fingerprint mismatch")
	}
	if !Verify(peerPub, reply.Payload, reply.Signature) {
		return nil, fmt.Errorf("signature verification failed")
	}

	// 6. Decapsulate.
	ss, err := dk.Decapsulate(reply.Payload)
	if err != nil {
		return nil, fmt.Errorf("mlkem decaps: %w", err)
	}

	// 7. Derive keys.
	keys := deriveKeys(ss)

	return newSecureConn(rw, keys, true), nil
}

// HandshakeResponder performs the responder side of the USCP handshake.
//
//  1. Receive HELLO
//  2. Verify peer fingerprint and signature
//  3. Encapsulate shared secret with peer's ek
//  4. Sign the ciphertext with our ML-DSA-87 identity
//  5. Send HELLO_REPLY
//  6. Derive session keys
func HandshakeResponder(rw io.ReadWriter, id *Identity, peerFP Fingerprint) (*SecureConn, error) {
	// 1. Receive HELLO.
	hello, err := unmarshalHello(rw.(io.Reader))
	if err != nil {
		return nil, fmt.Errorf("recv hello: %w", err)
	}
	if hello.Type != MsgHello {
		return nil, fmt.Errorf("expected HELLO (0x%02x), got 0x%02x", MsgHello, hello.Type)
	}

	// 2. Verify peer.
	peerPub, err := parseMDSA87PublicKey(hello.PublicKey)
	if err != nil {
		return nil, err
	}
	gotFP := computeFingerprintBytes(hello.PublicKey)
	if gotFP != peerFP {
		return nil, fmt.Errorf("fingerprint mismatch")
	}
	if !Verify(peerPub, hello.Payload, hello.Signature) {
		return nil, fmt.Errorf("signature verification failed")
	}

	// 3. Encapsulate.
	peerEK, err := mlkem.NewEncapsulationKey1024(hello.Payload)
	if err != nil {
		return nil, fmt.Errorf("parse mlkem ek: %w", err)
	}
	ss, ct := peerEK.Encapsulate()

	// 4. Sign ct.
	sig, err := id.Sign(ct)
	if err != nil {
		return nil, fmt.Errorf("sign ct: %w", err)
	}

	// 5. Send HELLO_REPLY.
	pubBytes, _ := id.PublicKey.MarshalBinary()
	reply := &helloMsg{
		Type:      MsgHelloReply,
		PublicKey: pubBytes,
		Payload:   ct,
		Signature: sig,
	}
	if _, err := rw.Write(reply.marshal()); err != nil {
		return nil, fmt.Errorf("send hello_reply: %w", err)
	}

	// 6. Derive keys.
	keys := deriveKeys(ss)

	return newSecureConn(rw, keys, false), nil
}

File: uniconn-js/core/src/secure/handshake.ts
import { MlKem1024 } from "mlkem";
import { blake3 } from "@noble/hashes/blake3.js";

import type { IConn } from "../conn.js";
import {
  computeFingerprint,
  type Fingerprint,
  type IIdentity,
  type VerifyFn,
} from "./identity.js";
import {
  MSG_HELLO,
  MSG_HELLO_REPLY,
  KDF_CONTEXT,
  KDF_OUTPUT_SIZE,
  NONCE_PREFIX_SIZE,
} from "./constants.js";
import {
  marshalHello,
  unmarshalHello,
  readFrame,
  writeFrame,
  type HelloMsg,
} from "./message.js";
import { SecureConn, type SessionKeys } from "./conn.js";

/** Derive session keys from ML-KEM shared secret using BLAKE3. */
function deriveKeys(sharedSecret: Uint8Array): SessionKeys {
  const ctxBytes = new TextEncoder().encode(KDF_CONTEXT);
  const combined = new Uint8Array(ctxBytes.length + sharedSecret.length);
  combined.set(ctxBytes);
  combined.set(sharedSecret, ctxBytes.length);
  const out = blake3(combined, { dkLen: KDF_OUTPUT_SIZE });

  return {
    initiatorKey: out.slice(0, 32),
    responderKey: out.slice(32, 64),
    initiatorNoncePfx: out.slice(64, 64 + NONCE_PREFIX_SIZE),
    responderNoncePfx: out.slice(64 + NONCE_PREFIX_SIZE, KDF_OUTPUT_SIZE),
  };
}

const kem = new MlKem1024();

/**
 * Perform the initiator side of the USCP handshake.
 *
 * @param conn  - Transport connection (TCP, WebSocket, WebTransport, etc.)
 * @param id    - Local identity (NodeIdentity or BrowserIdentity)
 * @param peerFP - Expected 64-byte fingerprint of the peer
 * @param verify - Platform-specific verify function (nodeVerify or browserVerify)
 */
export async function handshakeInitiator(
  conn: IConn,
  id: IIdentity,
  peerFP: Fingerprint,
  verify: VerifyFn,
): Promise<SecureConn> {
  // 1. Ephemeral ML-KEM-1024.
  const [ek, dk] = await kem.generateKeyPair();

  // 2. Sign raw ek.
  const sig = id.sign(ek);

  // 3. Send HELLO.
  const hello: HelloMsg = {
    type: MSG_HELLO,
    publicKey: id.publicKeyRaw,
    payload: ek,
    signature: sig,
  };
  await writeFrame(conn, marshalHello(hello));

  // 4. Receive HELLO_REPLY.
  const replyBody = await readFrame(conn);
  const reply = unmarshalHello(replyBody);
  if (reply.type !== MSG_HELLO_REPLY) {
    throw new Error(
      `expected HELLO_REPLY (0x${MSG_HELLO_REPLY.toString(16)}), got 0x${reply.type.toString(16)}`,
    );
  }

  // 5. Verify peer.
  const gotFP = computeFingerprint(reply.publicKey);
  if (!equalBytes(gotFP, peerFP)) {
    throw new Error("fingerprint mismatch");
  }
  if (!verify(reply.signature, reply.payload, reply.publicKey)) {
    throw new Error("signature verification failed");
  }

  // 6. Decapsulate.
  const sharedSecret = await kem.decap(reply.payload, dk);

  // 7. Derive keys.
  const keys = deriveKeys(sharedSecret);
  return new SecureConn(conn, keys, true);
}

/**
 * Perform the responder side of the USCP handshake.
 *
 * @param conn   - Transport connection
 * @param id     - Local identity
 * @param peerFP - Expected peer fingerprint
 * @param verify - Platform-specific verify function
 */
export async function handshakeResponder(
  conn: IConn,
  id: IIdentity,
  peerFP: Fingerprint,
  verify: VerifyFn,
): Promise<SecureConn> {
  // 1. Receive HELLO.
  const helloBody = await readFrame(conn);
  const hello = unmarshalHello(helloBody);
  if (hello.type !== MSG_HELLO) {
    throw new Error(
      `expected HELLO (0x${MSG_HELLO.toString(16)}), got 0x${hello.type.toString(16)}`,
    );
  }

  // 2. Verify peer.
  const gotFP = computeFingerprint(hello.publicKey);
  if (!equalBytes(gotFP, peerFP)) {
    throw new Error("fingerprint mismatch");
  }
  if (!verify(hello.signature, hello.payload, hello.publicKey)) {
    throw new Error("signature verification failed");
  }

  // 3. Encapsulate.
  const [ct, sharedSecret] = await kem.encap(hello.payload);

  // 4. Sign ct.
  const sig = id.sign(ct);

  // 5. Send HELLO_REPLY.
  const reply: HelloMsg = {
    type: MSG_HELLO_REPLY,
    publicKey: id.publicKeyRaw,
    payload: ct,
    signature: sig,
  };
  await writeFrame(conn, marshalHello(reply));

  // 6. Derive keys.
  const keys = deriveKeys(sharedSecret);
  return new SecureConn(conn, keys, false);
}

/** Constant-time byte comparison. */
function equalBytes(a: Uint8Array, b: Uint8Array): boolean {
  if (a.length !== b.length) return false;
  let diff = 0;
  for (let i = 0; i < a.length; i++) {
    diff |= a[i] ^ b[i];
  }
  return diff === 0;
}
